### ISSUE TITLE
buildbot master bytes issue with python3 when triggered via buildbot try

### ISSUE THREAD

**Issue by albertfong** [buildbot/buildbot#4488]
OS: ubuntu:bionic (docker base image)
Python: Python 3.6.7
Buildbot version: 1.7.0
Twisted version: 18.9.0

When submitting a build via buildbot try, the buildbot try command succeeds, and the build master receives the job, adds the buildset, and starts the build. However, the master errors out, leaving the builds in a bad state (UI shows started and running, but no progress is made).

build master logs suggest a bytes/string issue with the 'branch' property:

```master_1_29527f36cb6f | 2018-12-21 21:56:54+0000 [Broker,0,172.21.0.5] user ci requesting build on builders None
master_1_29527f36cb6f | 2018-12-21 21:56:54+0000 [-] added buildset 30 to database
master_1_29527f36cb6f | 2018-12-21 21:56:55+0000 [-] starting build <Build projects-linux-clang number:None results:success> using worker <WorkerForBuilder builder='projects-linux-clang' worker='clang' state=AVAILABLE>
master_1_29527f36cb6f | 2018-12-21 21:56:55+0000 [-] <Build projects-linux-clang number:None results:success>.startBuild
master_1_29527f36cb6f | 2018-12-21 21:56:55+0000 [-] from a running build; this is a serious error - please file a bug at http://buildbot.net
master_1_29527f36cb6f |         Traceback (most recent call last):
master_1_29527f36cb6f |           File "/usr/local/lib/python3.6/dist-packages/twisted/internet/defer.py", line 460, in callback
master_1_29527f36cb6f |             self._startRunCallbacks(result)
master_1_29527f36cb6f |           File "/usr/local/lib/python3.6/dist-packages/twisted/internet/defer.py", line 568, in _startRunCallbacks
master_1_29527f36cb6f |             self._runCallbacks()
master_1_29527f36cb6f |           File "/usr/local/lib/python3.6/dist-packages/twisted/internet/defer.py", line 654, in _runCallbacks
master_1_29527f36cb6f |             current.result = callback(current.result, *args, **kw)
master_1_29527f36cb6f |           File "/usr/local/lib/python3.6/dist-packages/twisted/internet/defer.py", line 1475, in gotResult
master_1_29527f36cb6f |             _inlineCallbacks(r, g, status)
master_1_29527f36cb6f |         --- <exception caught here> ---
master_1_29527f36cb6f |           File "/usr/local/lib/python3.6/dist-packages/twisted/internet/defer.py", line 1418, in _inlineCallbacks
master_1_29527f36cb6f |             result = g.send(result)
master_1_29527f36cb6f |           File "/usr/local/lib/python3.6/dist-packages/buildbot/process/build.py", line 309, in startBuild
master_1_29527f36cb6f |             self.setupOwnProperties()
master_1_29527f36cb6f |           File "/usr/local/lib/python3.6/dist-packages/buildbot/process/build.py", line 238, in setupOwnProperties
master_1_29527f36cb6f |             props.setProperty("branch", source.branch, "Build")
master_1_29527f36cb6f |           File "/usr/local/lib/python3.6/dist-packages/buildbot/process/properties.py", line 196, in setProperty
master_1_29527f36cb6f |             json.dumps(value)  # Let the exception propagate ...
master_1_29527f36cb6f |           File "/usr/lib/python3.6/json/__init__.py", line 231, in dumps
master_1_29527f36cb6f |             return _default_encoder.encode(obj)
master_1_29527f36cb6f |           File "/usr/lib/python3.6/json/encoder.py", line 199, in encode
master_1_29527f36cb6f |             chunks = self.iterencode(o, _one_shot=True)
master_1_29527f36cb6f |           File "/usr/lib/python3.6/json/encoder.py", line 257, in iterencode
master_1_29527f36cb6f |             return _iterencode(o, 0)
master_1_29527f36cb6f |           File "/usr/lib/python3.6/json/encoder.py", line 180, in default
master_1_29527f36cb6f |             o.__class__.__name__)
master_1_29527f36cb6f |         builtins.TypeError: Object of type 'bytes' is not JSON serializable
```

Here's the "buildbot try -n" output.

```
root@74cc4eb9177e:/source# buildbot try --connect=pb --username=<redacted> --passwd=<redacted> --master=master:9990 --vc=git --wait -n
using 'pb' connect method
job created
Job:
        Repository: /source
        Project:
        Branch: b'master'
        Revision: b'c6915c7046851d20940b12e968a19664f8d7c7d4'
        Builders: None
b' <redacted> '
job has been delivered
```

Something else I noticed. If I specify `--branch="master"` in the buildbot try invocation, then the type of Branch is a string. Probably another bug where the job property types differ if they're specified explicitly. The master still fails with a similar callstack as above, but instead with prop "revision".

```
master_1_6de1f963da00 |             props.setProperty("revision", source.revision, "Build")
master_1_6de1f963da00 |           File "/usr/local/lib/python3.6/dist-packages/buildbot/process/properties.py", line 196, in setProperty
master_1_6de1f963da00 |             json.dumps(value)  # Let the exception propagate ...
master_1_6de1f963da00 |           File "/usr/lib/python3.6/json/__init__.py", line 231, in dumps
master_1_6de1f963da00 |             return _default_encoder.encode(obj)
master_1_6de1f963da00 |           File "/usr/lib/python3.6/json/encoder.py", line 199, in encode
master_1_6de1f963da00 |             chunks = self.iterencode(o, _one_shot=True)
master_1_6de1f963da00 |           File "/usr/lib/python3.6/json/encoder.py", line 257, in iterencode
master_1_6de1f963da00 |             return _iterencode(o, 0)
master_1_6de1f963da00 |           File "/usr/lib/python3.6/json/encoder.py", line 180, in default
master_1_6de1f963da00 |             o.__class__.__name__)
master_1_6de1f963da00 |         builtins.TypeError: Object of type 'bytes' is not JSON serializable
```

If I downgrade my docker images to Python 2.7, then this scenario still works; I can submit buildbot try and trigger builds successfully.

**Comment by lantictac** [comment]
I'm also seeing this issue locally with the same stack trace.

OS: Windows-10-10.0.17763-SP0
Python: Python 3.7.0
Buildbot version: 1.7.0
Twisted version: 18.970

**Comment by lantictac** [comment]
I just tried some ugly hacking in tryclient.py to see if I could get it working again:

`
      class SourceStamp(object):

      def __init__(self, branch, revision, patch, repository=''):
          self.branch = branch.decode()
          self.revision = revision.decode()
          self.patch = (patch[0], patch[1].decode())
          self.repository = repository
`

This allows the try build submission to work again (although its obviously not a safe fix).

**Comment by rodrigc** [comment]
can you try this patch instead:

```
diff --git a/master/buildbot/process/properties.py b/master/buildbot/process/properties.py
index 543ddc026..bd70dad3d 100644
--- a/master/buildbot/process/properties.py
+++ b/master/buildbot/process/properties.py
@@ -193,6 +193,8 @@ class Properties(util.ComparableMixin):
 
     def setProperty(self, name, value, source, runtime=False):
         name = util.bytes2unicode(name)
+        if isinstance(value, bytes):
+            value = bytes2unicode(value)
         json.dumps(value)  # Let the exception propagate ...
         source = util.bytes2unicode(source)
```

**Comment by lantictac** [comment]
This patch appears to upset git:

```
BuildStep.failed; traceback follows
	Traceback (most recent call last):
	  File "c:\users\xx\appdata\local\programs\python\python37\lib\site-packages\twisted\internet\defer.py", line 501, in errback
	    self._startRunCallbacks(fail)
	  File "c:\users\xx\appdata\local\programs\python\python37\lib\site-packages\twisted\internet\defer.py", line 568, in _startRunCallbacks
	    self._runCallbacks()
	  File "c:\users\xx\appdata\local\programs\python\python37\lib\site-packages\twisted\internet\defer.py", line 654, in _runCallbacks
	    current.result = callback(current.result, *args, **kw)
	  File "c:\users\xx\appdata\local\programs\python\python37\lib\site-packages\twisted\internet\defer.py", line 1475, in gotResult
	    _inlineCallbacks(r, g, status)
	--- <exception caught here> ---
	  File "c:\users\xx\appdata\local\programs\python\python37\lib\site-packages\buildbot\process\buildstep.py", line 578, in startStep
	    self.results = yield self.run()
	  File "c:\users\xx\appdata\local\programs\python\python37\lib\site-packages\buildbot\process\buildstep.py", line 708, in run
	    results = yield self._start_deferred
	  File "c:\users\xx\appdata\local\programs\python\python37\lib\site-packages\buildbot\steps\source\git.py", line 199, in startVC
	    yield self._getAttrGroupMember('mode', self.mode)()
	  File "c:\users\xx\appdata\local\programs\python\python37\lib\site-packages\twisted\internet\defer.py", line 1418, in _inlineCallbacks
	    result = g.send(result)
	  File "c:\users\xx\appdata\local\programs\python\python37\lib\site-packages\buildbot\process\buildstep.py", line 708, in run
	    results = yield self._start_deferred
	  File "c:\users\xx\appdata\local\programs\python\python37\lib\site-packages\buildbot\steps\source\git.py", line 199, in startVC
	    yield self._getAttrGroupMember('mode', self.mode)()
	builtins.TypeError: startswith first arg must be bytes or a tuple of bytes, not str

```

**Comment by chrisspen** [comment]
I'm getting this as well, from somewhere in GitPoller, and it looks like a pretty common error caused by old Python2.7 code that hasn't been properly ported. Does Buildbot not fully support Python3 yet? The PyPI listing says it officially supports 3.6.

**Comment by sab24** [comment]
I made a pull request #5534 to fix this issue

**Comment by p12tic** [comment]
Fixed in #5623.
